### PR TITLE
feat: Added single_source_all_shortest_paths

### PR DESCRIPTION
**Pull request by gluonhiggs**
Resolve issue #1184 
Because The issue primarily focuses on single_source_all_shortest_paths, although it mentions an alternative function with 'or', so I have implemented only `single_source_all_shortest_paths`. The core logic is based on networkx implementation with some adaptations to `rustworkx`. The test cases are referred from `networkx` tests for `single_source_all_shortest_paths` also.

### CODE REVIEW

File: tests/graph/test_graph_single_source_all_shortest_paths.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            [
+                (self.disconnected_nodes[0], self.disconnected_nodes[1], 1),
+                (self.disconnected_nodes[1], self.disconnected_nodes[2], 1),
+                (self.disconnected_nodes[2], self.disconnected_nodes[3], 1),
+                (self.disconnected_nodes[3], self.disconnected_nodes[0], 1),
+            ]
+        )
+
+    def test_single_source_all_shortest_paths_cycle(self):
+        paths = rustworkx.graph_single_source_all_shortest_paths(self.cycle, self.cycle_nodes[0])
```
Comment (IvanIsCoding) [inline tests/graph/test_graph_single_source_all_shortest_paths.py:33]: Optional, but feel free to use `single_source_all_shortest_paths` directly. You can trust the dispatcher works once you added it

File: tests/graph/test_graph_single_source_all_shortest_paths.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
+import unittest
+import rustworkx
+
+
+class TestSingleSourceAllShortestPaths(unittest.TestCase):
```
Comment (IvanIsCoding) [inline tests/graph/test_graph_single_source_all_shortest_paths.py:5]: Add a test case with zero-weighted edges, and compare the output by filtering the output of https://www.rustworkx.org/apiref/rustworkx.all_simple_paths.html with Dijkstra's shortest path lenghts

File: tests/digraph/test_digraph_single_source_all_shortest_paths.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import unittest
+import rustworkx
+
+
+class TestDigraphSingleSourceAllShortestPaths(unittest.TestCase):
```
Comment (IvanIsCoding) [inline tests/digraph/test_digraph_single_source_all_shortest_paths.py:5]: Add a test case with zero-weighted edges, and compare the output by filtering the output of https://www.rustworkx.org/apiref/rustworkx.all_simple_paths.html with Dijkstra's shortest path lenghts

**Review by IvanIsCoding — COMMENTED** [review]
Excellent start! I left many comments, but this is already very close to getting merged.

* Add a dispatcher in https://github.com/Qiskit/rustworkx/blob/7840a2c606dcb3065e960bbd28df7437e59312c7/rustworkx/__init__.py#L413; you'll need to copy your docstring one more time; don't forget to add the warning there as well
* Add a release note to document your change and share with people your work! https://github.com/Qiskit/rustworkx/blob/main/CONTRIBUTING.md#release-notes

File: src/shortest_path/mod.rs [hunk 4]
```diff
@@ -324,6 +325,148 @@ pub fn digraph_all_shortest_paths(
         .collect())
 }
 
+/// Find all shortest paths from a single source node to all other nodes in an undirected graph
+///
+/// This function will generate all possible shortest paths from a source node to all other nodes
```
[the remark this answers is not in the document]
Comment (gluonhiggs) [reply src/shortest_path/mod.rs:327]: Done!

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+                }
+            }
+            if !predecessors.is_empty() {
+                pred.insert(v, predecessors);
+            }
+        }
+    }
+
+    // Memoized recursive function to compute all shortest paths
+    let mut memo: DictMap<G::NodeId, Vec<Vec<G::NodeId>>> = DictMap::new();
```
[the remark this answers is not in the document]
Comment (gluonhiggs) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs]: I analyzed which variables are potentially accessed frequently and changed only  `pred`. I no longer use `memo`, the reasons are given in the "trickiest part"  below :joy: .

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    fn all_paths<G: NodeIndexable>(
```
[the remark this answers is not in the document]
Comment (gluonhiggs) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs]: Well. I tried with ties breaking as you suggested, but it required an extra variable to track which target node we are handling, while we have to handle 0-weight edge cases. This is not "trickiest" part for me, it's extremely hard when I have to manage more constraints. So, I switch to backtracking, it's easier to work with. Do you want to stick with the previous version, or we open another issue to improve this?

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+    #[test]
+    fn test_single_source_all_shortest_paths_grid() {
+        let mut g = Graph::new_undirected();
+        let mut nodes = Vec::new();
+        for _ in 0..16 {
+            nodes.push(g.add_node(()));
+        }
+        // Add horizontal edges
+        for r in 0..4 {
```
[the remark this answers is not in the document]
Comment (gluonhiggs) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs]: Done!

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+            paths_list.sort_by(|p1, p2| p1.cmp(p2));
+        }
+        let mut actual_paths = paths.clone();
+        for paths_list in actual_paths.values_mut() {
+            paths_list.sort_by(|p1, p2| p1.cmp(p2));
+        }
+
+        assert_eq!(actual_paths, expected);
+    }
+}
```
[the remark this answers is not in the document]
Comment (gluonhiggs) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs:422]: Done!

File: src/shortest_path/mod.rs [hunk 9]
```diff
@@ -324,6 +325,148 @@ pub fn digraph_all_shortest_paths(
...
+    signature=(graph, source, weight_fn=None, default_weight=1.0),
+    text_signature = "(graph, source, /, weight_fn=None, default_weight=1.0)"
+)]
+pub fn graph_single_source_all_shortest_paths(
+    py: Python,
+    graph: &graph::PyGraph,
+    source: usize,
+    weight_fn: Option<PyObject>,
+    default_weight: f64,
+) -> PyResult<HashMap<usize, Vec<Vec<usize>>>> {
```
[the remark this answers is not in the document]
Comment (gluonhiggs) [reply src/shortest_path/mod.rs]: Done!

File: tests/graph/test_graph_single_source_all_shortest_paths.py (hunk 2 above)
Comment (gluonhiggs) [reply tests/graph/test_graph_single_source_all_shortest_paths.py:5]: Done!

File: tests/digraph/test_digraph_single_source_all_shortest_paths.py (hunk 3 above)
Comment (gluonhiggs) [reply tests/digraph/test_digraph_single_source_all_shortest_paths.py:5]: Done!

**Comment by IvanIsCoding** [comment]
I will give a review later, but my opinion is: handling the zero-weighted edges can be as complicated as as re-implementing all_simple_paths.

For now, just throw an error if there is a non-positive edge.

**Comment by gluonhiggs** [comment]
> I will give a review later, but my opinion is: handling the zero-weighted edges can be as complicated as as re-implementing all_simple_paths.
> 
> For now, just throw an error if there is a non-positive edge.

Actually I did a little bit deeper, that is to returns all shortest paths (with no repeated node). For example, in the test `test_single_source_all_shortest_paths_zero_weight`, with a graph of 4 nodes `a`,`b`,`c`, and `d` where edges `ab=bc=ca=0` and `cd=1`, all shortest path from `a` to `d` are `[[a,b,c,d],[a,c,d]]`.

File: src/shortest_path/mod.rs [hunk 10]
```diff
@@ -324,6 +328,159 @@ pub fn digraph_all_shortest_paths(
...
+    default_weight: f64,
+) -> PyResult<DictMap<usize, Vec<Vec<usize>>>> {
+    if source >= graph.node_count() {
+        return Err(PyIndexError::new_err("Source node index out of range"));
+    }
+
+    let start = NodeIndex::new(source);
+
+    // Define the cost function closure based on whether weight_fn is provided
+    let cost_fn_closure: CostFnClosure<'_> = if let Some(weight_fn) = weight_fn {
```
Comment (IvanIsCoding) [inline src/shortest_path/mod.rs]: Please refer to https://github.com/Qiskit/rustworkx/blob/7840a2c606dcb3065e960bbd28df7437e59312c7/src/shortest_path/all_pairs_bellman_ford.rs#L57 and https://github.com/Qiskit/rustworkx/blob/7840a2c606dcb3065e960bbd28df7437e59312c7/src/shortest_path/all_pairs_dijkstra.rs#L129 for how to get the edge weights with Python

- Because the algorithm might access the edge weights multiple times, storing them in Rust saves expensive trips to Python
- It also simplifies the signature, you shouldn't need `Box`

File: src/shortest_path/mod.rs [hunk 11]
```diff
@@ -324,6 +328,159 @@ pub fn digraph_all_shortest_paths(
...
+    as_undirected: bool,
+) -> PyResult<DictMap<usize, Vec<Vec<usize>>>> {
+    if source >= graph.node_count() {
+        return Err(PyIndexError::new_err("Source node index out of range"));
+    }
+
+    let start = NodeIndex::new(source);
+    let cost_fn = CostFn::try_from((weight_fn, default_weight))?;
+
+    // Define a closure that wraps CostFn to match the expected signature
```
Comment (IvanIsCoding) [inline src/shortest_path/mod.rs]: Same comment about edge weights

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+    {
+        if v == source {
+            let mut path = current_path.clone();
+            path.push(source);
+            path.reverse();
+            all_paths.push(path);
+            return;
+        }
+        for &p in &pred[graph.to_index(v)] {
+            if !current_path.contains(&p) {
```
Comment (IvanIsCoding) [inline rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs]: I think this check can become very expensive for large paths. In my opinion:
* Add https://github.com/Licenser/halfbrown as a dependency
* Make a "hashmap" of the indices with `halfbrown::HashMap<G::NodeId, ()>`
* Check with that; possibly pass it as an argument

Comment (IvanIsCoding) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs]: Maybe even better, just reuse a https://docs.rs/fixedbitset/latest/fixedbitset/ to keep track of what notes are currently in the path. That would be the fastest check, it wouldn't even require hashing

Comment (gluonhiggs) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs]: I also see that in the case of large paths, the time it takes to run is going to be long. 
I will look into this https://docs.rs/fixedbitset/latest/fixedbitset/.

File: src/shortest_path/mod.rs [hunk 13]
```diff
@@ -324,6 +322,189 @@ pub fn digraph_all_shortest_paths(
...
+            let cost_fn = CostFn::try_from((Some(weight_fn.clone()), default_weight))?;
+            cost_fn.call(py, edge_weight)
+        } else {
+            edge_weight.extract::<f64>(py)
+        }?;
+        edge_costs[pos] = Some(cost);
+    }
+
+    // Define cost function using precomputed costs
+    let cost_fn =
```
Comment (IvanIsCoding) [inline src/shortest_path/mod.rs:398]: I'd make this take `EdgeIndex` directly, just like in `all_pairs_dijkstra.rs`

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+type AllShortestPathsMap<N> = DictMap<N, Vec<Vec<N>>>;
+pub fn single_source_all_shortest_paths<G, F, K, E>(
+    graph: G,
+    source: G::NodeId,
+    mut edge_cost: F,
+) -> Result<AllShortestPathsMap<G::NodeId>, E>
+where
+    G: IntoEdgesDirected + Visitable + NodeIndexable + IntoNodeIdentifiers,
+    G::NodeId: Eq + Hash + Clone + Ord,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
```
Comment (IvanIsCoding) [inline rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs:16]: Make this take `EdgeIndex`, I think it would be more consistent with the rest of the library

File: src/shortest_path/mod.rs [hunk 15]
```diff
@@ -324,6 +322,189 @@ pub fn digraph_all_shortest_paths(
...
+    for edge_index in graph.graph.edge_indices() {
+        let pos = edge_index.index();
+        let edge_weight = graph.graph.edge_weight(edge_index).unwrap();
+        let cost = cost_fn.call(py, edge_weight)?;
+        edge_costs[pos] = Some(cost);
+    }
+
+    // Define cost function using precomputed costs
+    let cost_fn_closure =
+        move |edge: petgraph::stable_graph::EdgeReference<'_, Py<PyAny>>| -> Result<f64, PyErr> {
```
Comment (IvanIsCoding) [inline src/shortest_path/mod.rs:489]: Same comment about taking `EdgeIndex`

**Review by IvanIsCoding — COMMENTED** [review]
I left a minor comment about `EdgeIndex`, it shouldn't be too much to adddress.

After you add the type annotations for the universal method just like inhttps://github.com/Qiskit/rustworkx/blob/7840a2c606dcb3065e960bbd28df7437e59312c7/rustworkx/__init__.pyi#L634, CI should pass and I will merge this PR

File: src/shortest_path/mod.rs (hunk 13 above)
Comment (gluonhiggs) [reply src/shortest_path/mod.rs:392]: What is the benefit of doing this? Why?

Comment (gluonhiggs) [reply src/shortest_path/mod.rs:398]: I have updated this as you suggested.

File: rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs (hunk 14 above)
Comment (gluonhiggs) [reply rustworkx-core/src/shortest_path/single_source_all_shortest_paths.rs:16]: I'm not quite clear about this. Do you mean we have to make `FnMut(EdgeIndex)`? If so, I don't find any code in `rustworkx` that is similar to that.

**Comment by gluonhiggs** [comment]
Thank you @IvanIsCoding
